haskell/hsc2hs.bzl

```diff
@@ -17,19 +19,61 @@ def hsc_to_hs(ctx):
   Returns:
     list of File: New Haskell source files to use.
   """
+  ghc_defs_dump = _make_ghc_defs_dump(ctx)
   sources = []
   for f in ctx.files.srcs:
     if f.extension == "hsc":
-      sources.append(_process_hsc_file(ctx, f))
+      sources.append(_process_hsc_file(ctx, ghc_defs_dump, f))
     else:
       sources.append(f)
   return sources
 
-def _process_hsc_file(ctx, hsc_file):
+def _make_ghc_defs_dump(ctx):
+  """Generate a file containing GHC default pre-processor definitions.
+
+  Args:
+    ctx: Rule context.
+
+  Returns:
+    File: The file with GHC definitions.
+  """
+  raw_filename = "ghc-defs-dump-{0}-{1}.hs".format(ctx.attr.name, ctx.attr.version)
+  dummy_src = ctx.actions.declare_file(raw_filename)
+  ghc_defs_dump0 = ctx.actions.declare_file(paths.replace_extension(raw_filename, ".hspp"))
+  ctx.actions.write(dummy_src, "")
+  args0 = ctx.actions.args()
+  args0.add([
+    "-E",
+    "-optP-dM",
+    "-cpp",
+    dummy_src.path,
+  ])
+  ctx.actions.run(
+    inputs = depset([dummy_src]),
+    outputs = [ghc_defs_dump0],
+    executable = get_compiler(ctx),
+    arguments = [args0],
+  )
+  ghc_defs_dump = ctx.actions.declare_file(paths.replace_extension(raw_filename, ".h"))
+
+  ctx.actions.run_shell(
+    inputs = depset([ghc_defs_dump0]),
+    outputs = [ghc_defs_dump],
+    command = "tail -n +2 $1 > $2",
+    arguments = [
+      ghc_defs_dump0.path,
+      ghc_defs_dump.path,
+    ],
+  )
+
+  return ghc_defs_dump
```

Comment:
This probably needs some filtering:

```
bazel-out/k8-fastbuild/bin/external/haskell_old_time/ghc-defs-dump-old-time-1.0.0.h:17:0: warning: "__STDC_HOSTED__" redefined
 #define __STDC_HOSTED__ 1
```

And more importantly, all kinds of things break. I haven't investigated why yet, but I guess some of these defines shouldn't be defined:
```
In file included from /usr/include/x86_64-linux-gnu/sys/time.h:25:0,
                 from external/haskell_old_time/include/HsTime.h:19,
                 from Time.hsc:111:
/usr/include/time.h:75:1: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'typedef'
 typedef __time_t time_t;
 ^~~~~~~
/usr/include/time.h:76:1: error: unknown type name '__END_NAMESPACE_STD'
 __END_NAMESPACE_STD
 ^~~~~~~~~~~~~~~~~~~
```